### DISCUSSION TITLE
Websocket recovery

### DISCUSSION THREAD

**Discussion by f82a** [miguelgrinberg/microdot#210] [Q&A]
Hello
the websocket connection closes if it is not used after a certain time. If I initialize a new websocket connection in Javascript (let socket = new WebSocket('ws://' + location.host + '/echo');) the message arrives with a new websocket instance on the Micropyten server. However, the response to the client is then lost - probably because of the new websocket instance that was initialized by the client (?)

Is this behavior normal? How can I initiate a bidirectional working websocket connection from the client - without reloading the page again?

Thanks for a hint!

@app.route('/echo')
@with_websocket
async def echo(request, ws):
        data = await ws.receive() 
        await ws.send(data)

**Comment by miguelgrinberg** [comment]
The WebSocket implementation in Microdot does not have any timeouts. But it is fairly common for web servers, proxy servers or clients to include timeouts after some idle time. Are you sure these timeouts aren't happening somewhere else? In any case, you can send a dummy message every once in a while to prevent the timeout.

**Comment by f82a** [comment]
Thanks for your reply!
Yes, I can prevent a timeout with a periodic polling. The problem is that in my use case the browser is often in the background and so the polling also fails. The advantage would be that when a WebSocket connection is closed, it is simply reinitialized by the client. As described, this only works to a limited extent. The server receives such a request from the client, but the response does not reach the client (unknown instance?). Here is the Javascript function i use for re-opening the WegSocket:

function send_data(msg) {
    if (socket.readyState === WebSocket.OPEN) {  
        socket.send(msg);
    } else {
        socket = new WebSocket('ws://' + location.host + '/echo'); 
        socket.addEventListener('open', function(event) {
            socket.send(msg);
        });
    };
}

**Reply by miguelgrinberg** [reply]
I do not understand what you are proposing. If your connection is timing out there is no way to reactivate it, you need to make a new connection, which in turn will invoke a new route with a new websocket instance.
